File: src/quotes.ts
import { DataClient, nowMs } from "./predexon.js";

export type TopOfBook = {
  venue: "polymarket" | "limitless";
  bestBid: number;
  bestAsk: number;
  bidSize: number;
  askSize: number;
};

export const spread = (q: TopOfBook): number => Math.max(0, q.bestAsk - q.bestBid);
export const mid = (q: TopOfBook): number => (q.bestAsk + q.bestBid) / 2;
export const topDepth = (q: TopOfBook): number => Math.min(q.bidSize, q.askSize);

const extractBest = (levels: Array<{ price: number; size: number }>, side: "bid" | "ask"): { price: number; size: number } => {
  if (!levels.length) return { price: 0, size: 0 };
  const prices = levels.map((l) => Number(l.price)).filter((p) => Number.isFinite(p));
  if (!prices.length) return { price: 0, size: 0 };
  const bestPrice = side === "bid" ? Math.max(...prices) : Math.min(...prices);
  const best = levels.find((l) => Number(l.price) === bestPrice);
  return { price: bestPrice, size: Number(best?.size ?? 0) };
};

export const quotePolymarketToken = async (
  data: DataClient,
  tokenId: string,
  lookbackMs = 5 * 60 * 1000
): Promise<TopOfBook | null> => {
  const endMs = nowMs();
  const startMs = endMs - lookbackMs;
  const payload = await data.polymarketOrderbooks({ tokenId, startMs, endMs, limit: 1 });
  const snaps = (payload?.snapshots ?? []) as any[];
  if (!snaps.length) return null;
  const s = snaps[snaps.length - 1];
  const bids = (s?.bids ?? []) as any[];
  const asks = (s?.asks ?? []) as any[];
  const b = extractBest(bids, "bid");
  const a = extractBest(asks, "ask");
  if (b.price <= 0 || a.price <= 0) return null;
  return { venue: "polymarket", bestBid: b.price, bestAsk: a.price, bidSize: b.size, askSize: a.size };
};

export const quoteLimitlessMarket = async (
  data: DataClient,
  marketSlug: string,
  side: "yes" | "no" = "yes",
  lookbackMs = 5 * 60 * 1000
): Promise<TopOfBook | null> => {
  const endMs = nowMs();
  const startMs = endMs - lookbackMs;
  const payload = await data.limitlessOrderbooks({ marketSlug, startMs, endMs, limit: 1 });
  const snaps = (payload?.snapshots ?? []) as any[];
  if (!snaps.length) return null;
  const s = snaps[snaps.length - 1];
  const bids = (s?.bids ?? []) as any[];
  const asks = (s?.asks ?? []) as any[];
  const b = extractBest(bids, "bid");
  const a = extractBest(asks, "ask");
  if (b.price <= 0 || a.price <= 0) return null;
  if (side === "yes") {
    return { venue: "limitless", bestBid: b.price, bestAsk: a.price, bidSize: b.size, askSize: a.size };
  }
  const bestBid = 1 - a.price;
  const bestAsk = 1 - b.price;
  const bidSize = a.size;
  const askSize = b.size;
  return { venue: "limitless", bestBid, bestAsk, bidSize, askSize };
};

File: predexonooor/quotes.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

from .predexon import DataClient, now_ms


@dataclass(frozen=True)
class TopOfBook:
    venue: str
    best_bid: float
    best_ask: float
    bid_size: float
    ask_size: float

    @property
    def spread(self) -> float:
        return max(0.0, self.best_ask - self.best_bid)

    @property
    def mid(self) -> float:
        return (self.best_ask + self.best_bid) / 2.0


def _extract_best(levels: list[dict[str, Any]], *, side: Literal["bid", "ask"]) -> tuple[float, float]:
    if not levels:
        return 0.0, 0.0
    prices = [float(l["price"]) for l in levels if "price" in l]
    if not prices:
        return 0.0, 0.0
    if side == "bid":
        best_price = max(prices)
    else:
        best_price = min(prices)
    best_levels = [l for l in levels if float(l.get("price", -1)) == best_price]
    size = float(best_levels[0].get("size", 0.0)) if best_levels else 0.0
    return best_price, size


def quote_polymarket_token(data: DataClient, *, token_id: str, lookback_ms: int = 5 * 60 * 1000) -> TopOfBook | None:
    end_ms = now_ms()
    start_ms = end_ms - lookback_ms
    payload = data.polymarket_orderbooks(token_id=token_id, start_ms=start_ms, end_ms=end_ms, limit=1)
    snapshots = payload.get("snapshots") or []
    if not snapshots:
        return None
    snap = snapshots[-1]
    bids = snap.get("bids") or []
    asks = snap.get("asks") or []
    best_bid, bid_size = _extract_best(bids, side="bid")
    best_ask, ask_size = _extract_best(asks, side="ask")
    if best_bid <= 0.0 or best_ask <= 0.0:
        return None
    return TopOfBook(venue="polymarket", best_bid=best_bid, best_ask=best_ask, bid_size=bid_size, ask_size=ask_size)


def quote_limitless_market(data: DataClient, *, market_slug: str, lookback_ms: int = 5 * 60 * 1000) -> TopOfBook | None:
    end_ms = now_ms()
    start_ms = end_ms - lookback_ms
    payload = data.limitless_orderbooks(market_slug=market_slug, start_ms=start_ms, end_ms=end_ms, limit=1)
    snapshots = payload.get("snapshots") or []
    if not snapshots:
        return None
    snap = snapshots[-1]
    bids = snap.get("bids") or []
    asks = snap.get("asks") or []
    best_bid, bid_size = _extract_best(bids, side="bid")
    best_ask, ask_size = _extract_best(asks, side="ask")
    if best_bid <= 0.0 or best_ask <= 0.0:
        return None
    return TopOfBook(venue="limitless", best_bid=best_bid, best_ask=best_ask, bid_size=bid_size, ask_size=ask_size)
